samples/databricks-cicd/README.md

```diff
@@ -0,0 +1,70 @@
+# Overview
+
+__What does this sample demonstrate:__
+
+This sample demonstrates how to run unit tests on Python code using [Databricks Connect](https://docs.microsoft.com/en-us/azure/databricks/dev-tools/databricks-connect), run integration tests on Databricks notebooks, and deploy a Databricks job, all in [GitHub Actions](https://docs.github.com/en/actions/learn-github-actions).
+
+This sample contains a single notebook [main_notebook.py](src/main_notebook.py) which calls [some_function](src/dbcicdlib/some_func.py) in a Python module. We run [unit test on some_function](test/unittests/test_some_func.py) and [integration test on main_notebook.py](test/run_notebook_tests.sh).
+
+This sample also shows how to [update the job if a job with the same name already exists](../../.github/workflows/databricks-cicd.yml#L136). Without this, if the CD pipeline publishes a job, it's a new job with a new job id even if another job with the same name already exists.
+
+Another small feature used in this sample is to use a [pre-commit config hook](.pre-commit-config.yaml) to run [detect-secrets](https://github.com/Yelp/detect-secrets) to prevent secrets from being checked into code.
+
+__What doesn't this sample demonstrate:__
+
+Both the notebook and the library code are bare minimum sample code, not meant to represent real world business logic.
+
+## How it works
+
+Databricks Connect enables IDE such as VSCode and [Databricks Connect Cli](https://docs.microsoft.com/en-us/azure/databricks/dev-tools/databricks-connect#step-1-install-the-client) to access Databricks cluster from the local dev machine or build agent. Unit tests should have no external dependencies such as dbfs, instead, they should take dataframes as input/output parameters. Once Databricks Connect is set up, you can run unit tests using PyTest as usual, either in command line, or in VSCode. You can also debug Python scripts or unit tests in VSCode. If the target Databricks cluster is not running, it will be automatically started.
```

Comment:
Just for my understanding, we are doing unit tests, but they are run within a/the Databricks cluster? So in a sense I have the one dependency to that cluster. Within the cluster we don't have any further external dependencies (storage, db etc)?

Reply:
Yes, that is correct. Basically you'll need a Spark engine to run anything Spark. 

One way to do it is to install Spark in a local environment or building agent (can be a container) and run code there. You do need to make sure the Spark version you run and the configurations are same as what you'd run in Databricks. Even so, there are still differences such as dbfs. 

The other way to do it is to run code on Databricks cluster. That's the approach taken by Databricks Connect and Nutter. The downside is that you have this external dependency, the upside is you can get the exact same environment you'd run in prod without having to create your own Spark environment.

Reply:
Ok makes sense, in the end we will always have dependencies, e.g. the build agent and the tools need to run it. Was a big caught in unit testing philosophy. For the developers productivity however I think unit test ideally can be executed quickly any time without the need for spinning up a environment, but that can be done with a small Databricks cluster being constantly available.